subsys/bluetooth/controller/ecdh.c

```diff
@@ -0,0 +1,339 @@
+#include <zephyr.h>
+#include <sys/atomic.h>
+
+#include <sys/byteorder.h>
+
+#include <bluetooth/hci.h>
+#include <bluetooth/buf.h>
+#include <bluetooth/hci_err.h>
+#include <drivers/bluetooth/hci_driver.h>
+
+#include <mpsl/mpsl_work.h>
+
+#include "ecdh.h"
+
+#define BT_DBG_ENABLED IS_ENABLED(CONFIG_BT_DEBUG_HCI_DRIVER)
+#define LOG_MODULE_NAME bt_ecdh
+#include "common/log.h"
+
+static struct {
+	uint8_t private_key_be[32];
+
+	union {
+		uint8_t public_key_be[64];
+		uint8_t dhkey_be[32];
+	};
+} ecdh;
+
+enum {
+	GEN_PUBLIC_KEY  = BIT(0),
+	GEN_DHKEY       = BIT(1),
+	GEN_DHKEY_DEBUG = BIT(2),
+
+};
+atomic_t cmd;
+
+/* based on Core Specification 4.2 Vol 3. Part H 2.3.5.6.1 */
+static const uint8_t debug_private_key_be[32] = {
+	0x3f, 0x49, 0xf6, 0xd4, 0xa3, 0xc5, 0x5f, 0x38,
+	0x74, 0xc9, 0xb3, 0xe3, 0xd2, 0x10, 0x3f, 0x50,
+	0x4a, 0xff, 0x60, 0x7b, 0xeb, 0x40, 0xb7, 0x99,
+	0x58, 0x99, 0xb8, 0xa6, 0xcd, 0x3c, 0x1a, 0xbd,
+};
+
+#if defined(CONFIG_BT_CTLR_ECDH_LIB_OBERON)
+#include <ocrypto_ecdh_p256.h>
+
+static uint8_t public_key(void)
+{
+	int err;
+
+	do {
+		err = bt_rand(ecdh.private_key_be, 32);
+		if (err) {
+			return BT_HCI_ERR_UNSPECIFIED;
+		}
+
+		if (!memcmp(ecdh.private_key_be, debug_private_key_be, 32)) {
+			err = -1;
+			continue;
+		}
```

Comment:
In order to make debug keys possible I made a change in upstream, it will be pulled in with the upmerge.
The host then uses the V2 of the gen dhkey command to generate it using the public key.
The host knows the debug public key, so there is no need to generate it.